core/psh/pshapp/pshapp.c

```diff
@@ -0,0 +1,1179 @@
+		return -ENOMEM;
+	}
+
+	for (;;) {
+		write(STDOUT_FILENO, "\r\033[0J", 5);
+		write(STDOUT_FILENO, PROMPT, sizeof(PROMPT) - 1);
+
+		if ((n = psh_readcmd(&orig, cmdhist, &cmd)) < 0) {
+			err = n;
+			break;
+		}
+
+		if ((err = psh_parsecmd(cmd, &argc, &argv)) < 0) {
+			free(cmd);
+			if (err == -EINVAL)
+				continue;
+			break;
+		}
+
+		/* Select command history entry */
+		if (cmdhist->he != cmdhist->hb) {
+			entry = &cmdhist->entries[(cmdhist->he) ? cmdhist->he - 1 : HISTSZ - 1];
+			if ((n == entry->n) && !memcmp(cmd, entry->cmd, n)) {
+				cmdhist->he = (cmdhist->he) ? cmdhist->he - 1 : HISTSZ - 1;
+				free(entry->cmd);
+			}
+			else {
+				entry = cmdhist->entries + cmdhist->he;
+			}
+		}
+		else {
+			entry = cmdhist->entries + cmdhist->he;
+		}
+
+		/* Update command history */
+		entry->cmd = cmd;
+		entry->n = n;
+		if ((cmdhist->he = (cmdhist->he + 1) % HISTSZ) == cmdhist->hb) {
+			free(cmdhist->entries[cmdhist->hb].cmd);
+			cmdhist->entries[cmdhist->hb].cmd = NULL;
+			cmdhist->hb = (cmdhist->hb + 1) % HISTSZ;
+		}
+
+		/* Clear signals */
+		psh_common.sigint = 0;
+		psh_common.sigquit = 0;
+		psh_common.sigstop = 0;
+
+		/* Reset getopt */
+		optind = 1;
+
+		/* Find and run */
+		if ((app = psh_findapp(argv[0])) == NULL && argv[0][0] == '/')
+			app = psh_findapp("/");
+
+		if (app != NULL)
+			app->run(argc, argv);
+		else
+			printf("Unknown command!\n");
+			
+		free(argv);
+		fflush(NULL);
+
+		if (pshapp_common.isrunpsh == 0)
+			break;
+	}
+
+	/* Free command history */
+	for (; cmdhist->hb != cmdhist->he; cmdhist->hb = (cmdhist->hb + 1) % HISTSZ)
+		free(cmdhist->entries[cmdhist->hb].cmd);
+
+	free(cmdhist);
+
+	return err;
+}
+
+
+int psh_pshappexit(int argc, char **argv)
+{
+	if (pshapp_common.isrunpsh == 1)
+		pshapp_common.isrunpsh = 0;
+	return 0;
+}
+
+
+void psh_pshappexitinfo(void)
+{
+	printf("exits shell");
+}
+
+
+int psh_pshapp(int argc, char **argv)
+{
+	char *path = NULL;
+	int c;
+
+	/* Run shell script */
+	if (argc > 1) {
+		while ((c = getopt(argc, argv, "i:h")) != -1) {
+			switch (c) {
+			case 'i':
+				path = optarg;
+				break;
+
+			case 'h':
+			default:
+				printf("usage: %s [options] [script path] or no args to run shell interactively\n", argv[0]);
+				printf("  -i <script path>:  selects psh script to execute\n");
+				printf("  -h:                shows this help message\n");
+				return EOK;
+			}
+		}
+
+		if (optind < argc)
+			path = argv[optind];
+
+		if (path != NULL)
+			return psh_runscript(path);
+	}
+	/* Run shell interactively */
+	else {
+		if (pshapp_common.isrunpsh == 0) {
+			pshapp_common.isrunpsh = 1;
+			psh_run(1);
+			pshapp_common.isrunpsh = 0;
+		}
+	}
+
+	return EOK;
+}
+
+
+int psh_pshapplogin(int argc, char **argv)
+{
+	const psh_appentry_t *loginapp = psh_findapp("login");
+	while (1) {
+		if (pshapp_common.isrunpsh != 0)
+			return -EACCES;
+		if ( pshapp_common.isrunpsh == 0 && (loginapp == NULL || loginapp->run(argc, argv) == 1 )) {
```

Comment:
`isrunpsh` is a flag that ensures only one active command line (`pshapp`) in scope of one `psh` process. 

The double-check is redundant provided that login procedure (`loginapp->run()`) is fast (low probability of flag change during login). However the login procedure can last a long time (user doing nothing when presented with login prompt) so my intention was to just double check it for sure.

But it is good you pointed this out because if above logic should work the order of argument in `if` must be changed to check `isrunpsh` after `loginapp->run()`.